Dead partition because of duplicated keys in Column Family

### <!-- Component -->

<!-- Zeebe- -->

### <!-- Affected version -->

<!-- -8.8 -->

### <!-- Severity -->

<!-- Critical- -->

### <!-- Likelihood -->

<!-- Unknown_ -->

### Description

A partition may fail to start if a duplicated mapping is found. As of now it occurred a couple of times in `zeebe-dev` envinronment:
- [Key DbCompositeKey{first=oid, second=${INITIAL_CLAIM_VALUE}} in ColumnFamily MAPPINGS already exists](https://cloudlogging.app.goo.gl/tH9aFuk4mf23Tr836)
- [Key DbCompositeKey{first=DbCompositeKey{first=io.camunda.zeebe.db.impl.DbEnumValue@3ed636dc, second=oid-${INITIAL_CLAIM_VALUE}}, second=DbCompositeKey{first=io.camunda.zeebe.db.impl.DbEnumValue@51004013, second=admin}} in ColumnFamily RELATIONS_BY_ENTITY already exists](https://cloudlogging.app.goo.gl/iPQZNodv3hDY8n2c6)

The cause of the issue may be that `oid-${INITIAL_CLAIM_VALUE}` is not getting substituted correctly

### Steps to reproduce

Not sure


### Current behavior

The partition is marked as dead because it fails to start

### Expected behavior

Partition should be able to start correctly

### Environment

SaaS

### Version

- Zeebe SNAPSHOT


### Rootcause

_No response_

### Solution ideas

_No response_

### Dev -> QA handover




### Links

<details>
<summary> complete log line </summary>
<pre>
{
  "jsonPayload": {
    "message": "Actor StreamProcessor-1 failed in phase STARTED.",
    "exception": "java.lang.RuntimeException: Failed to replay batch at 'LoggedEvent [position=24207, key=2251799813685249, timestamp=1745920741489, sourceEventPosition=24206] RecordMetadata{recordType=EVENT, valueType=ROLE, intent=ENTITY_ADDED, authorization=AuthInfo{format=UNKNOWN, authData=, claims={}}}'\n\tat io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$4(ReplayStateMachine.java:170) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:86) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:43) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:125) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:127) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:104) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:224) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\nCaused by: io.camunda.zeebe.db.ZeebeDbInconsistentException: Key DbCompositeKey{first=DbCompositeKey{first=io.camunda.zeebe.db.impl.DbEnumValue@3ed636dc, second=oid-${INITIAL_CLAIM_VALUE}}, second=DbCompositeKey{first=io.camunda.zeebe.db.impl.DbEnumValue@51004013, second=admin}} in ColumnFamily RELATIONS_BY_ENTITY already exists\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.assertKeyDoesNotExist(TransactionalColumnFamily.java:316) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$insert$0(TransactionalColumnFamily.java:87) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$ensureInOpenTransaction$19(TransactionalColumnFamily.java:344) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:31) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.ensureInOpenTransaction(TransactionalColumnFamily.java:343) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.insert(TransactionalColumnFamily.java:82) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.engine.state.authorization.DbMembershipState.insertRelation(DbMembershipState.java:55) ~[zeebe-workflow-engine-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.engine.state.appliers.RoleEntityAddedApplier.applyState(RoleEntityAddedApplier.java:29) ~[zeebe-workflow-engine-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.engine.state.appliers.RoleEntityAddedApplier.applyState(RoleEntityAddedApplier.java:17) ~[zeebe-workflow-engine-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:669) ~[zeebe-workflow-engine-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.engine.Engine.replay(Engine.java:107) ~[zeebe-workflow-engine-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.stream.impl.ReplayStateMachine.replayEvent(ReplayStateMachine.java:237) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat java.base/java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]\n\tat io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$tryToReplayBatch$5(ReplayStateMachine.java:210) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:109) ~[zeebe-db-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.stream.impl.ReplayStateMachine.tryToReplayBatch(ReplayStateMachine.java:208) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$3(ReplayStateMachine.java:164) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:28) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.retry.RecoverableRetryStrategy.run(RecoverableRetryStrategy.java:48) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\tat io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:84) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]\n\t... 5 more\n",
    "context": {
      "partitionId": "1",
      "threadId": 49,
      "threadPriority": 5,
      "loggerName": "io.camunda.zeebe.logstreams",
      "threadName": "zb-actors-0",
      "actor-name": "StreamProcessor-1",
      "actor-scheduler": "Broker-2"
    },
  },
  "resource": {
    "type": "k8s_container",
    "labels": {
      "project_id": "camunda-saas-dev",
      "pod_name": "zeebe-2",
      "container_name": "zeebe",
      "namespace_name": "7c2e6d37-5b3c-4022-8fa8-d8299fa71711-zeebe",
      "location": "europe-north1",
      "cluster_name": "worker-2"
    }
  },
  "timestamp": "2025-04-29T09:59:01.522652851Z",
  "severity": "ERROR",
  "logName": "projects/camunda-saas-dev/logs/stdout",
  "sourceLocation": {
    "file": "StreamProcessor.java",
    "line": "389",
    "function": "onFailure"
  },
  "receiveTimestamp": "2025-04-29T09:59:04.164310879Z",
  "errorGroups": [
    {
      "id": "CPDM9-CV9Nvk3wE"
    }
  ]
}
</pre>
</details>

--- a/authentication/src/main/java/io/camunda/authentication/CamundaOAuthPrincipalService.java
+++ b/authentication/src/main/java/io/camunda/authentication/CamundaOAuthPrincipalService.java
@@ -11,6 +11,7 @@ import io.camunda.authentication.entity.AuthenticationContext;
 import io.camunda.authentication.entity.OAuthContext;
 import io.camunda.search.entities.GroupEntity;
 import io.camunda.search.entities.MappingEntity;
+import io.camunda.search.entities.RoleEntity;
 import io.camunda.security.configuration.SecurityConfiguration;
 import io.camunda.security.entity.AuthenticationMethod;
 import io.camunda.service.AuthorizationServices;
@@ -78,9 +79,7 @@ public class CamundaOAuthPrincipalService {
             getUsernameFromClaims(claims),
             assignedRoles,
             authorizationServices.getAuthorizedApplications(
-                Stream.concat(
-                        assignedRoles.stream().map(r -> r.roleKey().toString()),
-                        mappingIds.stream())
+                Stream.concat(assignedRoles.stream().map(RoleEntity::roleId), mappingIds.stream())
                     .collect(Collectors.toSet())),
             tenantServices.getTenantsByMemberIds(mappingIds).stream()
                 .map(TenantDTO::fromEntity)

--- a/authentication/src/test/java/io/camunda/authentication/CamundaOAuthPrincipalServiceTest.java
+++ b/authentication/src/test/java/io/camunda/authentication/CamundaOAuthPrincipalServiceTest.java
@@ -23,8 +23,6 @@ import io.camunda.service.GroupServices;
 import io.camunda.service.MappingServices;
 import io.camunda.service.RoleServices;
 import io.camunda.service.TenantServices;
-import java.time.Instant;
-import java.time.temporal.ChronoUnit;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -34,10 +32,6 @@ import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 
 public class CamundaOAuthPrincipalServiceTest {
-  private static final String REGISTRATION_ID = "test";
-  private static final String TOKEN_VALUE = "{}";
-  private static final Instant TOKEN_ISSUED_AT = Instant.now().truncatedTo(ChronoUnit.SECONDS);
-  private static final Instant TOKEN_EXPIRES_AT = TOKEN_ISSUED_AT.plus(1, ChronoUnit.DAYS);
 
   private CamundaOAuthPrincipalService camundaOAuthPrincipalService;
 

--- a/authentication/src/test/java/io/camunda/authentication/CamundaOAuthPrincipalServiceTest.java
+++ b/authentication/src/test/java/io/camunda/authentication/CamundaOAuthPrincipalServiceTest.java
@@ -86,7 +86,7 @@ public class CamundaOAuthPrincipalServiceTest {
     final var roleR1 = new RoleEntity(8L, "roleR1", "Role R1");
     when(roleServices.getRolesByMemberIds(Set.of("test-id", "test-id-2")))
         .thenReturn(List.of(roleR1));
-    when(authorizationServices.getAuthorizedApplications(Set.of("test-id", "test-id-2", "8")))
+    when(authorizationServices.getAuthorizedApplications(Set.of("test-id", "test-id-2", "roleR1")))
         .thenReturn(List.of("*"));
 
     // when
@@ -122,7 +122,7 @@ public class CamundaOAuthPrincipalServiceTest {
     final var roleR1 = new RoleEntity(10L, "roleR1", "Role R1");
     when(roleServices.getRolesByMemberIds(Set.of("map-1", "map-2"))).thenReturn(List.of(roleR1));
 
-    when(authorizationServices.getAuthorizedApplications(Set.of("map-1", "map-2", "10")))
+    when(authorizationServices.getAuthorizedApplications(Set.of("map-1", "map-2", "roleR1")))
         .thenReturn(List.of("app-1", "app-2"));
 
     // when

